zerver/lib/bugdown/__init__.py

```diff
@@ -698,6 +701,28 @@ def handleMatch(self, match):
         img.set('alt', email)
         return img
 
+class Timestamp(markdown.inlinepatterns.Pattern):
+    def handleMatch(self, match):
+        # type: (Match[Text]) -> Optional[Element]
+        span = markdown.util.etree.Element('span')
+        span.set('class', 'timestamp')
+        anchor = markdown.util.etree.SubElement(span, 'span')
+        try:
+            timestamp = dateutil.parser.parse(match.group('time'))
+            span.set('value', timestamp.strftime("%s"))
+            anchor.set('value', timestamp.strftime("%s"))
+        except ValueError:
+            try:
+                timestamp = datetime.fromtimestamp(float(match.group('time')))
+                span.set('value', timestamp.strftime("%s"))
+                anchor.set('value', timestamp.strftime("%s"))
+            except ValueError:
+                pass
+        # Set text to initial input, so even if parsing fails, the data remains intact.
+        span.text = ":clock4:"
+        anchor.text = match.group('time')
+        return span
```

Comment:
I think we should probably change the markup for these to something simpler (without the emoji), e.g.:

`<span class=\"timestamp\" value=\"1496701800\">Jun 5th 2017, 10:30PM</span>`

And then if we want the emoji, we'd include it via a `:before` entry in the CSS (though I'd be happy with skipping it for now and planning to add it or something else in as we get feedback).  Not having extra structure in the markup for this will make this syntax much cleaner and more maintainable across future design changes.

Reply:
After thinking about this a bit more, I think the raw text should probably be in ISO format (or something similarly standard that encodes the timezone as well as the date); that seems like a better experience for any clients that don't properly support the rewriting on this feature.